wavesexchange_repos/src/circuit_breaker/mod.rs

```diff
@@ -0,0 +1,249 @@
+pub mod config;
+
+pub use config::Config;
+use wavesexchange_log::debug;
+
+use std::{
+    future::Future,
+    mem::drop,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tokio::sync::RwLock;
+
+pub trait FallibleDataSource {
+    type Error;
+
+    fn is_countable_err(err: &Self::Error) -> bool;
+
+    fn fallback(&self, elapsed_ms: u128, err_count: usize) -> Self::Error {
+        panic!(
+            "CircuitBreaker panicked after {err_count} errors in a row happened in {elapsed_ms}ms"
+        )
+    }
+}
+
+pub trait DataSrcInitFn<S: FallibleDataSource>:
+    Fn() -> Result<S, S::Error> + Send + Sync + 'static
```

Comment:
If possible, in all APIs the `FnMut` should be used, because it is less restrictive for the user of this API.